File: Nissan-vercel-main/apps/api/app/routers/company_summary.py
"""Company Summary API (Phase 3) — see docs/planner/03_COMPANY_SUMMARY.md.

A standalone vertical (reads website_extractions read-only; never modifies
agents/website_extraction/ or agents/context_planner/), so it gets its own
router — mirrors app/routers/website_extraction.py's shape. Unlike Phase 2,
POST here is synchronous — one bounded Groq call, not a multi-page crawl.
"""
from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from agents.company_summary.service import (
    ExtractionNotEligible,
    create_summary,
    get_summary,
    list_summaries,
)

router = APIRouter()


# ── Request / Response models ────────────────────────────────────────────────

class CreateSummaryRequest(BaseModel):
    tenant_id: str
    extraction_id: str


class SummaryResponse(BaseModel):
    summary_id: str
    tenant_id: str
    extraction_id: str
    context_id: str
    status: str
    company_name: str | None = None
    website: str | None = None
    region: str | None = None
    industry: str | None = None
    products: list[str] = []
    services: list[str] = []
    description: str | None = None
    verdict: str | None = None
    errors: list[str] = []
    created_at: str | None = None
    updated_at: str | None = None


def _dict_to_response(d: dict) -> SummaryResponse:
    return SummaryResponse(
        summary_id=d["id"],
        tenant_id=d["tenant_id"],
        extraction_id=d["extraction_id"],
        context_id=d["context_id"],
        status=d.get("status", "pending"),
        company_name=d.get("company_name"),
        website=d.get("website"),
        region=d.get("region"),
        industry=d.get("industry"),
        products=d.get("products") or [],
        services=d.get("services") or [],
        description=d.get("description"),
        verdict=d.get("verdict"),
        errors=d.get("errors") or [],
        created_at=d.get("created_at"),
        updated_at=d.get("updated_at"),
    )


# ── Endpoints ─────────────────────────────────────────────────────────────────

@router.post("/summaries", response_model=SummaryResponse)
async def create_summary_endpoint(body: CreateSummaryRequest) -> SummaryResponse:
    """Synchronous: validates the referenced extraction is ready, runs the
    graph inline, and returns the completed (ready or failed) summary in one
    response — no polling needed, unlike Phase 2's crawl."""
    try:
        result = await create_summary(body.tenant_id, body.extraction_id)
    except ExtractionNotEligible as e:
        raise HTTPException(status_code=400, detail=str(e))
    return _dict_to_response(result)


@router.get("/summaries/{summary_id}", response_model=SummaryResponse)
async def get_summary_endpoint(summary_id: str, tenant_id: str) -> SummaryResponse:
    row = await get_summary(summary_id, tenant_id)
    if not row:
        raise HTTPException(status_code=404, detail=f"Summary {summary_id} not found")
    return _dict_to_response(row)


@router.get("/summaries", response_model=list[SummaryResponse])
async def list_summaries_endpoint(
    tenant_id: str, extraction_id: str | None = None, context_id: str | None = None,
    status: str | None = None, limit: int = 50,
) -> list[SummaryResponse]:
    rows = await list_summaries(tenant_id, extraction_id, context_id, status, limit)
    return [_dict_to_response(r) for r in rows]

File: Nissan-vercel-main/apps/web/src/lib/company-summary.ts
/**
 * Company Summary API client — Server-side functions (Phase 3).
 * See docs/planner/03_COMPANY_SUMMARY.md.
 */
import { createServerFn } from '@tanstack/react-start'
import { getSupabaseServerClient } from './supabase.server'

const AGENT_BASE = (import.meta.env.VITE_AGENT_API_URL as string | undefined) ?? 'http://localhost:8000'

const DEMO_TENANT_ID = '11111111-1111-1111-1111-111111111111'

async function tenantId(supabase: ReturnType<typeof getSupabaseServerClient>) {
  const {
    data: { user },
  } = await supabase.auth.getUser()
  if (!user) return DEMO_TENANT_ID
  const { data } = await supabase.from('users').select('tenant_id').eq('id', user.id).single()
  return (data?.tenant_id as string) ?? DEMO_TENANT_ID
}

export interface SummaryResult {
  summary_id: string
  tenant_id: string
  extraction_id: string
  context_id: string
  status: 'pending' | 'ready' | 'failed'
  company_name: string | null
  website: string | null
  region: string | null
  industry: string | null
  products: Array<string>
  services: Array<string>
  description: string | null
  verdict: string | null
  errors: Array<string>
  created_at: string | null
  updated_at: string | null
}

// Synchronous — one bounded Groq call, unlike the async extraction endpoint.
export const createSummary = createServerFn({ method: 'POST' })
  .validator((d: { extraction_id: string }) => d)
  .handler(async ({ data }): Promise<SummaryResult> => {
    const supabase = getSupabaseServerClient()
    const tenant_id = await tenantId(supabase)

    const res = await fetch(`${AGENT_BASE}/company-summary/summaries`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ tenant_id, extraction_id: data.extraction_id }),
    })
    if (!res.ok) {
      const err = (await res.json().catch(() => ({ detail: 'Failed to generate summary' }))) as { detail?: string }
      throw new Error(typeof err.detail === 'string' ? err.detail : 'Failed to generate summary')
    }
    return res.json() as Promise<SummaryResult>
  })

export const getSummary = createServerFn({ method: 'GET' })
  .validator((d: { summary_id: string }) => d)
  .handler(async ({ data }): Promise<SummaryResult | null> => {
    const supabase = getSupabaseServerClient()
    const tenant_id = await tenantId(supabase)

    const res = await fetch(`${AGENT_BASE}/company-summary/summaries/${data.summary_id}?tenant_id=${tenant_id}`)
    if (!res.ok) return null
    return res.json() as Promise<SummaryResult>
  })

export const listSummaries = createServerFn({ method: 'GET' })
  .validator((d: { context_id?: string; extraction_id?: string; limit?: number } | undefined) => d ?? {})
  .handler(async ({ data }): Promise<Array<SummaryResult>> => {
    const supabase = getSupabaseServerClient()
    const tenant_id = await tenantId(supabase)

    const params = new URLSearchParams({ tenant_id, limit: String(data.limit ?? 20) })
    if (data.context_id) params.set('context_id', data.context_id)
    if (data.extraction_id) params.set('extraction_id', data.extraction_id)
    const res = await fetch(`${AGENT_BASE}/company-summary/summaries?${params}`)
    if (!res.ok) return []
    return res.json() as Promise<Array<SummaryResult>>
  })
